### ISSUE TITLE
NCCL Blocking Send/Recv are Non-blocking in practice

### ISSUE THREAD

**Issue by andoorve** [pytorch/pytorch#129341]
### 🐛 Describe the bug

Blocking send and recv is non-blocking in-practice for the NCCL backend. The first send/recv pair is blocking, possibly due to  warmup with NCCLUniqueID exchange but not the subsequent recvs.

```
import time
import os
import torch
import torch.distributed as dist
import torch.multiprocessing as mp

def run(rank, size):
    """ Distributed function to be implemented later. """
    device = torch.device(f"cuda:{rank}")
    if rank == 0:
        tens = torch.ones([33, 4096], dtype=torch.bfloat16, device=device)
        dist.send(tens, dst=1)
        tens *= 2
        dist.send(tens, dst=1)
        print(f'SEND COMPLETED {time.ctime()}', flush=True)
    else:
        tens = torch.empty([33, 4096], dtype=torch.bfloat16, device=device)
        dist.recv(tens, src=0)
        tens = torch.empty([33, 4096], dtype=torch.bfloat16, device=device)
        time.sleep(10)
        print(f'ABOUT TO BEGIN RECV {time.ctime()}', flush=True)
        dist.recv(tens, src=0)
        print (f'{tens}', flush=True)

def init_process(rank, size, fn, backend='nccl'):
    """ Initialize the distributed environment. """
    os.environ['MASTER_ADDR'] = '127.0.0.1'
    os.environ['MASTER_PORT'] = '29501'
    dist.init_process_group(backend, rank=rank, world_size=size)
    fn(rank, size)


if __name__ == "__main__":
    size = 2
    processes = []
    mp.set_start_method("spawn")
    for rank in range(size):
        p = mp.Process(target=init_process, args=(rank, size, run, 'nccl'))
        p.start()
        processes.append(p)

    for p in processes:
        p.join()
```

```
SEND COMPLETED Mon Jun 24 02:20:36 2024
ABOUT TO BEGIN RECV Mon Jun 24 02:20:46 2024
tensor([[2., 2., 2.,  ..., 2., 2., 2.],
        [2., 2., 2.,  ..., 2., 2., 2.],
        [2., 2., 2.,  ..., 2., 2., 2.],
        ...,
        [2., 2., 2.,  ..., 2., 2., 2.],
        [2., 2., 2.,  ..., 2., 2., 2.],
        [2., 2., 2.,  ..., 2., 2., 2.]], device='cuda:1', dtype=torch.bfloat16)
```

Change device to `"cpu"` and backend to `'gloo'` to observe that the same thing doesn't happen with that backend and send/recvs are coordinated as expected.

cc: @kwen2501 @wconstab @youkaichao 

### Versions

```
Collecting environment information...
PyTorch version: 2.3.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.29.0
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.5.0-1016-gcp-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.3.107
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA L4
GPU 1: NVIDIA L4
GPU 2: NVIDIA L4
GPU 3: NVIDIA L4

Nvidia driver version: 545.23.08
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.0.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.0.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.0.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.0.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.0.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.0.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.0.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.0.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             96
On-line CPU(s) list:                0-95
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) CPU @ 2.20GHz
CPU family:                         6
Model:                              85
Thread(s) per core:                 2
Core(s) per socket:                 24
Socket(s):                          2
Stepping:                           7
BogoMIPS:                           4400.42
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat avx512_vnni md_clear arch_capabilities
Hypervisor vendor:                  KVM
Virtualization type:                full
L1d cache:                          1.5 MiB (48 instances)
L1i cache:                          1.5 MiB (48 instances)
L2 cache:                           48 MiB (48 instances)
L3 cache:                           77 MiB (2 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-23,48-71
NUMA node1 CPU(s):                  24-47,72-95
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Retbleed:             Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced / Automatic IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Mitigation; Clear CPU buffers; SMT Host state unknown

Versions of relevant libraries:
[pip3] mypy==1.9.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] torch==2.3.0
[pip3] triton==2.3.0
[conda] Could not collect
```

**Comment by youkaichao** [comment]
Indeed there are two lazy things happening here:

1. pytorch NCCL backend creates NCCL communicators lazily (at the first call of NCCL operations)
2. cuda module loading is lazy, every cuda kernel call is synchronized at the first launch.

**Comment by andoorve** [comment]
Hi, any update on this one?

**Comment by garrett361** [comment]
@andoorve I'm being slow: what's the issue your reproducer shows? I tried running on CPU and I think I'm seeing the same behavior as above. I'm missing what is unexpected here and would like to understand.

**Comment by andoorve** [comment]
@garrett361 

The issue is basically what you see in this print here:

```
SEND COMPLETED Mon Jun 24 02:20:36 2024
ABOUT TO BEGIN RECV Mon Jun 24 02:20:46 2024
```

The timestamps for NCCL are 10 seconds apart, indicating we completed the send call before the receive call. There should be an asymmetry with GLOO on your end, you should see both at the exact same timestamp.

It seems like this though is expected behaviour and a consequence of CUDA operations being launched asynchronously and documented here: https://pytorch.org/docs/stable/distributed.html#synchronous-and-asynchronous-collective-operations
although it might make sense to document this difference directly in the send/recv section.

Relevant lines:

> When the function returns, it is guaranteed that the collective operation is performed. In the case of CUDA operations, it is not guaranteed that the CUDA operation is completed, since CUDA operations are asynchronous. For CPU collectives, any further function calls utilizing the output of the collective call will behave as expected.

**Comment by Tomorrowdawn** [comment]
After searching online for a while, I came across this issue. There should be three devices involved here: CPU, GPU, and remote GPU. The synchronization described in send/recv operations should refer to that between the GPU and the remote GPU, not the GPU and the CPU.

I also think this is well worth documenting; otherwise, it can easily cause confusion.
